blog/content/post/2020-06-30-trenchboot-grub-cbfs.md

```diff
@@ -0,0 +1,575 @@
+
+1. Install `xxd` tool.
+
+    > This tool will be used during `landing-zone` build, so it must be done
+    earlier.
+
+    ```
+    $ cd ~/nixpkgs
+    $ nix-build -A unixtools.xxd
+    (...)
+    /nix/store/2q94zc1agpkvchxxnx6pwy1v6rpdqzdx-xxd-1003.1-2008
+    ```
+
+    Last line points to directory where package is installed. Copy `/bin/xxd`
+    file to one of the `PATH` directory.
+
+    ```
+    $ echo $PATH
+    /run/wrappers/bin:/root/.nix-profile/bin:/etc/profiles/per-user/root/bin:/nix/var/nix/profiles/default/bin
+    $ cp /nix/store/2q94zc1agpkvchxxnx6pwy1v6rpdqzdx-xxd-1003.1-2008/bin/xxd /run/wrappers/bin/
+    ```
+
+    > Adding `xxd` binary to PATH is necessary, so landing-zone compilation
+    process will end successfully.
+
+2. Update `nixpkgs` repository.
+
+    It is assumed that you have already downloaded custom
+    [3mdeb/nixpkg](https://github.com/3mdeb/nixpkgs) repository to your NixOS.
+    If yes, then go to `nixpkgs` directory and pull latest changes.
+
+    ```
+    $ cd ~/nixpkgs
+    $ git pull
+    ```
+
+    Change branch to `tpm_event_log`
+
+    ```
+    $ git checkout tpm_event_log
+    ```
+
+3. Rebuild NixOS.
+
+    ```
+    $ sudo nixos-rebuild switch -I nixpkgs=~/nixpkgs
+    $ reboot
+    ```
+
+    There are changes in `grub-tb` and `landing-zone` packages. Moreover there
+    is new package `cbmem-tb` added. It contains `cbmem` utility, which is a
+    tool for gathering early logs. It lets reading TPM event logs from OS level.
+
+4. Replace `/boot/lz_header` with just built new one.
+
+    ```
+    $ cd /nix/store
+    $ ls | grep -i 'landing-zone-0.6.0'
+    bqi0phr2rvwqlzfr3qj5117arxjhlbil-landing-zone-0.6.0
+    nw8rsi5jfx9zikv06dhjxc6a5219xr20-landing-zone-0.6.0.drv
+    ```
+
+    >`landing-zone-0.6.0` without any extension is directory you looking for.
+
+    ```
+    $ cp /nix/store/bqi0phr2rvwqlzfr3qj5117arxjhlbil-landing-zone-0.6.0/lz_header.bin /boot/lz_header
+    ```
+
+5. Reboot platform and boot to NixOS Secure Launch.
+
+6. Install `cbmem-tb` package manually.
+
+    ```
+    $ cd ~/nixpkgs
+    $ nix-build -A cbmem-tb
+    (...)
+    /nix/store/8y25mfqw0igqa5yfpvrks0nvr5wah5kn-cbmem-4.12
+    ```
+
+7. Copy `cbmem` utility to home directory.
+
+    ```
+    $ cd /nix/store/
+    $ ls | grep -i 'cbmem-4.12'
+    8y25mfqw0igqa5yfpvrks0nvr5wah5kn-cbmem-4.12
+    b4pds6l5fbkxdykkf2248c4mfnhf5ll5-cbmem-4.12.drv
+    plm3jyg54hjyiy3zldclx83k14i34lpq-cbmem-4.12.drv
+    pzifnqhgk0xarm3xrgkn07s9l572aw5p-cbmem-4.12.lock
+    ```
+
+    Search for entry with `cbmem-4.12` without any extension. In above case
+    `cbmem` is installed in
+    `/nix/store/8y25mfqw0igqa5yfpvrks0nvr5wah5kn-cbmem-4.12` directory.
+
+    ```
+    $ cp /nix/store/8y25mfqw0igqa5yfpvrks0nvr5wah5kn-cbmem-4.12/sbin/cbmem ~/
+    ```
+
+7. Go to home directory and use `cbmem` to read TPM event log.
+
+    ```
+    $ cd ~
+    $ ./cbmem -d
+    DRTM TPM2 log:
```

Comment:
Ohh this log needs formatting. There is a lack of new lines:
```
        Specification: 2.00
        Platform class: PC Client
        No vendor information provided
    DRTM TPM2 log entry 1:
        PCR: 17
        Event type: Unknown (0x600)
        Digests:
                 SHA1: 2400e5bdfbaa8cfc42eae13d9b742b89d0ba35b4
                 SHA256: b65067767baf988b18e3a83410f90f055a4dcd59509b1ab6b17e18926ad8de82
        Event data not provided
    DRTM TPM2 log entry 2:
        PCR: 17
        Event type: Unknown (0x601)
        Digests:
                 SHA1: ecad3658a0cda535a8db50c207d726d2dac46509
                 SHA256: f76f571cb78beddba64ceab81b679ab3328ec15e2c4ff49f9fff625c300dc5a1
        Event data: Kernel
```
I will add to the develop branch what is needed

Reply:
@piotr-kleins the output should be formatted like above. I have pushed a commit that fixes it, you may try and verify